internals-js/src/definitions.ts

```diff
@@ -495,11 +496,37 @@ export class Extension<TElement extends ExtendableElement> {
   }
 }
 
+type UnappliedDirective = {
+  nameOrDef: DirectiveDefinition<Record<string, any>> | string,
+  args: Record<string, any>,
+  extension?: Extension<any>,
+  directive: DirectiveNode,
+};
+
 // TODO: ideally, we should hide the ctor of this class as we rely in places on the fact the no-one external defines new implementations.
 export abstract class SchemaElement<TOwnType extends SchemaElement<any, TParent>, TParent extends SchemaElement<any, any> | Schema> extends Element<TParent> {
   protected readonly _appliedDirectives: Directive<TOwnType>[] = [];
+  protected readonly _unappliedDirectives: UnappliedDirective[] = [];
+
   description?: string;
 
+  addUnappliedDirective({ nameOrDef, args, extension, directive }: UnappliedDirective) {
+    this._unappliedDirectives.push({
+      nameOrDef,
+      args: args ?? {},
+      extension,
+      directive,
+    });
+  }
+
+  processUnappliedDirectives() {
+    for (const { nameOrDef, args, extension, directive } of this._unappliedDirectives) {
+      const d = this.applyDirective(nameOrDef, args);
+      d.setOfExtension(extension);
+      d.sourceAST = directive;
+    }
+  }
```

Comment:
**thought** We _may_ want to clear the `_unappliedDirectives` after they've been processed